akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderImpl.scala

```diff
@@ -45,68 +42,79 @@ import org.apache.kafka.common.TopicPartition
+    numPartitionsF.failed.foreach(_ => metadataClient.stop())
     numPartitionsF.map { numPartitions =>
-      Consumer
-        .plainPartitionedManualOffsetSource(settings, subscription, getOffsetsOnAssign(readOffsets))
+      _source(readOffsets)
         .flatMapMerge(numPartitions, {
           case (_, partitionedSource) => partitionedSource
         })
         .watchTermination()(Keep.right)
         .mapMaterializedValue { terminated =>
-          terminated.onComplete(_ => stopMetadataClient())
+          terminated.onComplete(_ => metadataClient.stop())
         }
     }
   }
 
-  override def extractOffset(record: ConsumerRecord[K, V]): MergeableOffset[Long] = {
-    val key = record.topic() + "-" + record.partition()
-    MergeableOffset(Map(key -> record.offset()))
+  override def extractOffset(record: ConsumerRecord[K, V]): GroupOffsets = GroupOffsets(record)
+
+  override def verifyOffset(offsets: GroupOffsets): OffsetVerification = {
+    if ((assignedPartitions.get() -- offsets.partitions) == EmptyTps)
+      SkipOffset("The offset contains Kafka topic partitions that were revoked or lost in a previous rebalance")
+    else
+      Success
   }
 
   private def getOffsetsOnAssign(readOffsets: ReadOffsets): Set[TopicPartition] => Future[Map[TopicPartition, Long]] =
     (assignedTps: Set[TopicPartition]) =>
       readOffsets()
         .flatMap {
-          case Some(mergeableOffsets) =>
-            Future.successful(mergeableOffsets.entries.flatMap {
-              case (surrogateProjectionKey, offset) =>
-                val tp = parseProjectionKey(surrogateProjectionKey)
+          case Some(groupOffsets) =>
+            Future.successful(groupOffsets.entries.flatMap {
+              case (topicPartitionKey, offset) =>
+                val tp = topicPartitionKey.tp
                 if (assignedTps.contains(tp)) Map(tp -> offset)
                 else Map.empty
             })
           case None => metadataClient.getBeginningOffsets(assignedTps)
         }
         .recover {
-          case NonFatal(ex) => throw new RuntimeException("External offsets could not be retrieved", ex)
+          case ex => throw new RuntimeException("External offsets could not be retrieved", ex)
         }
 
-  private def parseProjectionKey(surrogateProjectionKey: String): TopicPartition = {
-    surrogateProjectionKey match {
-      case RegexTp(topic, partition) => new TopicPartition(topic, partition.toInt)
-      case _ =>
-        throw new IllegalArgumentException(
-          s"Row entry name (${surrogateProjectionKey}) must match pattern: ${RegexTp.pattern.toString}")
-    }
+  private class ProjectionPartitionHandler extends PartitionAssignmentHandler {
+    override def onRevoke(revokedTps: Set[TopicPartition], consumer: RestrictedConsumer): Unit =
+      removeFromAssigned(revokedTps)
+
+    override def onAssign(assignedTps: Set[TopicPartition], consumer: RestrictedConsumer): Unit =
+      assignedPartitions.set(assignedTps)
```

Comment:
shouldn't this use `assignedPartitions.compareAndSet` and retry if another thread updated at the same time?

Reply:
Hmm.  Actually this should never be updated at the same time since it's all driven by the polling thread in `KafkaConsumerActor`, and each callback is executed serially.  I used a threadsafe datastructure here because the read case can happen in another thread, but maybe that's not necessary?

Reply:
Ok.  Thanks.